src/dag/lazy-store.ts

```diff
@@ -0,0 +1,350 @@
+  }
+
+  close(): void {
+    if (!this._closed) {
+      this._release();
+      this._sourceRead?.then(read => read.close());
+      this._closed = true;
+    }
+  }
+
+  get closed(): boolean {
+    return this._closed;
+  }
+
+  private async _getSourceRead(): Promise<Read> {
+    if (!this._sourceRead) {
+      this._sourceRead = this._sourceStore.read();
+    }
+    return this._sourceRead;
+  }
+}
+
+export class LazyWrite extends LazyRead implements Write {
+  private readonly _refCounts: Map<Hash, number>;
+  private readonly _chunkHasher: ChunkHasher;
+  protected readonly _pendingHeadChanges = new Map<string, HeadChange>();
+  protected readonly _pendingChunks = new Map<Hash, Chunk | undefined>();
+
+  constructor(
+    heads: Map<string, Hash>,
+    tempChunks: Map<Hash, Chunk>,
+    sourceChunksCache: ChunksCache,
+    sourceStore: Store,
+    refCounts: Map<Hash, number>,
+    release: () => void,
+    chunkHasher: ChunkHasher,
+    assertValidHash: (hash: Hash) => void,
+  ) {
+    super(
+      heads,
+      tempChunks,
+      sourceChunksCache,
+      sourceStore,
+      release,
+      assertValidHash,
+    );
+    this._refCounts = refCounts;
+    this._chunkHasher = chunkHasher;
+  }
+  createChunk = <V extends ReadonlyJSONValue>(
+    data: V,
+    refs: readonly Hash[],
+  ): Chunk<V> => createChunk(data, refs, this._chunkHasher);
+
+  async putChunk(c: Chunk): Promise<void> {
+    const {hash, meta} = c;
+    // TODO not sure this is really needed in this store
+    this.assertValidHash(hash);
+    if (meta.length > 0) {
+      for (const h of meta) {
+        this.assertValidHash(h);
+      }
+    }
+    this._pendingChunks.set(hash, c);
+  }
+  async setHead(name: string, hash: Hash): Promise<void> {
+    await this._setHead(name, hash);
+  }
+  async removeHead(name: string): Promise<void> {
+    await this._setHead(name, undefined);
+  }
+
+  private async _setHead(name: string, hash: Hash | undefined): Promise<void> {
+    const oldHash = await this.getHead(name);
+    const v = this._pendingHeadChanges.get(name);
+    if (v === undefined) {
+      this._pendingHeadChanges.set(name, {new: hash, old: oldHash});
+    } else {
+      // Keep old if existing
+      v.new = hash;
+    }
+  }
+
+  override async hasChunk(hash: Hash): Promise<boolean> {
+    return this._pendingChunks.get(hash) !== undefined || super.hasChunk(hash);
+  }
+
+  async getChunk(hash: Hash): Promise<Chunk | undefined> {
+    if (this._pendingChunks.has(hash)) {
+      return this._pendingChunks.get(hash);
+    }
+    return super.getChunk(hash);
+  }
+
+  async getHead(name: string): Promise<Hash | undefined> {
+    const headChange = this._pendingHeadChanges.get(name);
+    if (headChange) {
+      return headChange.new;
+    }
+    return super.getHead(name);
+  }
+
+  async commit(): Promise<void> {
+    const pendingRefCountUpdates = new Map<Hash, number | undefined>();
+    const garbargeCollector = new GarbargeCollector(
```

Comment:
Nit: Find this hard to read without names for the callbacks. Consider defining a `GarbageCollector::Host` interface or type to make this more self-documenting.  Can still pass in an inline definition of it just as you are now (which is nice).

Reply:
Or use a delegate... which could be `this`.